ilc/client/ClientRouter.js

```diff
@@ -48,48 +51,72 @@ export default class ClientRouter {
         return deepmerge(appProps, routeProps);
     }
 
-    #setInitialRoutes = () => {
+    #setInitialRoutes = (state) => {
         // we should respect base tag for cached pages
-        let path;
         const base = document.querySelector('base');
-
         if (base) {
             const a = document.createElement('a');
             a.href = base.getAttribute('href');
-            path = a.pathname + a.search;
+            this.#currentRoute = this.#router.match(a.pathname + a.search);
 
             base.remove();
             this.#logger.warn(
                 'ILC: <base> tag was used only for initial rendering and removed afterwards.\n' +
                 'Currently, ILC does not support it fully.\n' +
                 'Please open an issue if you need this functionality.'
             );
+        } else if (state.forceSpecialRoute === '404') {
+            this.#currentRoute = this.#router.matchSpecial(this.#getCurrUrl(), state.forceSpecialRoute);
         } else {
-            path = this.#location.pathname + this.#location.search;
+            this.#currentRoute = this.#router.match(this.#getCurrUrl());
         }
 
-        this.#currentRoute = this.#router.match(path);
         this.#prevRoute = this.#currentRoute;
     };
 
     #addEventListeners = () => {
-        window.addEventListener('single-spa:before-routing-event', this.#onSingleSpaRoutingEvents);
+        this.#windowEventHandlers['single-spa:before-routing-event'] = this.#onSingleSpaRoutingEvents;
+        this.#windowEventHandlers['ilc:404'] = this.#onSpecialRouteTrigger(404);
+
+        for (let key in this.#windowEventHandlers) {
+            if (!this.#windowEventHandlers.hasOwnProperty(key)) {
+                continue;
+            }
+
+            window.addEventListener(key, this.#windowEventHandlers[key]);
+        }
+
         document.addEventListener('click', this.#onClickLink);
     };
 
     removeEventListeners() {
-        window.removeEventListener('single-spa:before-routing-event', this.#onSingleSpaRoutingEvents);
+        for (let key in this.#windowEventHandlers) {
+            if (!this.#windowEventHandlers.hasOwnProperty(key)) {
+                continue;
+            }
+
+            window.removeEventListener(key, this.#windowEventHandlers[key]);
+        }
+        this.#windowEventHandlers = {};
+
         document.removeEventListener('click', this.#onClickLink);
     }
 
     #onSingleSpaRoutingEvents = () => {
         this.#prevRoute = this.#currentRoute;
 
+        const newUrl = this.#getCurrUrl();
+        if (this.#forceSpecialRoute !== null && this.#forceSpecialRoute.url === newUrl) {
+            this.#currentRoute = this.#router.matchSpecial(newUrl, this.#forceSpecialRoute.id);
```

Comment:
Can we reset "#forceSpecialRoute" immediately after using it here?
Why we should do it in separate `else if`?